src/transformers/trainer_pt_utils.py

```diff
@@ -55,6 +55,17 @@
 logger = logging.get_logger(__name__)
 
 
+def get_dataloader_sampler(dataloader):
+    # after accelerate.prepare function the wraped dataloader.sampler will be SequentialSampler instead of RandomSampler
+    if hasattr(dataloader, "sampler") and isinstance(dataloader.sampler, RandomSampler):
+        return dataloader.sampler
+    elif hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
+        return get_dataloader_sampler(dataloader.batch_sampler)
+    elif hasattr(dataloader, "sampler"):
+        return dataloader.sampler
+    return None
```

Comment:
@dumpmemory Could you please explain why you keep adding the `isinstance` check into this function? In the third logic check `elif  hasattr(dataloader, "sampler"):` means we will still be returning SequentialSampler. As I mentioned in the previous review, it's not needed. Neither is the final `return None` - this already happens in the function. 

If it's the case that when SequentialSampler is set, then the dataloader can also have a batch_sampler that is None and a sampler, we can just change the order of checks.

```suggestion
def get_dataloader_sampler(dataloader):
    if hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
        return get_dataloader_sampler(dataloader.batch_sampler)
    elif hasattr(dataloader, "sampler"):
        return dataloader.sampler
```